dm/syncer/error.go

```diff
@@ -61,6 +64,103 @@ func isDropColumnWithIndexError(err error) bool {
 			strings.Contains(mysqlErr.Message, "with tidb_enable_change_multi_schema is disable"))
 }
 
+// GetDDLStatusFromTiDB() retrieves the synchronizing status of DDL from TiDB
+// hence here db should be TiDB database.
+func GetDDLStatusFromTiDB(ctx context.Context, db *sql.DB, ddl string, createTime int64) (string, error) {
+	rowNum := 10
+	count := 0
+	for {
+		showJobs := fmt.Sprintf("ADMIN SHOW DDL JOBS %d", rowNum)
+		// every attempt try 10 history jobs
+		rows, err := db.QueryContext(ctx, showJobs)
+		if err != nil {
+			rows.Close()
+			return "", err
+		}
+
+		columns, err := rows.Columns()
+		if err != nil {
+			rows.Close()
+			return "", err
+		}
+
+		values := make([]sql.RawBytes, len(columns))
+		scanArgs := make([]interface{}, len(values))
+		for i := range values {
+			scanArgs[i] = &values[i]
+		}
+
+		// skip the lines that have been checked
+		skipNum := 0
+		if rowNum > 10 {
+			for rows.Next() {
+				skipNum++
+				if skipNum >= rowNum-10 {
+					break
+				}
+			}
+		}
+		for rows.Next() {
+			err = rows.Scan(scanArgs...)
+			if err != nil {
+				rows.Close()
+				return "", err
+			}
+
+			ddlCreateTimeStr := string(values[8])
+			var ddlCreateTimeParse time.Time
+			ddlCreateTimeParse, err = time.Parse("2006-01-02 15:04:05", ddlCreateTimeStr)
+			if err != nil {
+				rows.Close()
+				return "", err
+			}
+			ddlCreateTime := ddlCreateTimeParse.Unix()
+
+			// ddlCreateTime and createTime are both based on timezone of downstream
+			if ddlCreateTime >= createTime {
+				var jobID int
+				jobID, err = strconv.Atoi(string(values[0]))
+				if err != nil {
+					rows.Close()
+					return "", err
+				}
+
+				offset := rowNum + count - 10
+				for {
+					var DDLJob string
+					var jobIDForLimit int
+					showJob := fmt.Sprintf("ADMIN SHOW DDL JOB QUERIES LIMIT 1 OFFSET %d", offset)
+					err = db.QueryRowContext(ctx, showJob).Scan(&jobIDForLimit, &DDLJob)
+					if err != nil {
+						rows.Close()
+						return "", err
+					}
+					if jobID == jobIDForLimit && ddl == DDLJob {
+						rows.Close()
+						return string(values[11]), err
+					}
+					if jobIDForLimit <= jobID {
+						break
+					}
+					offset++
+				}
+				count++
+			} else {
+				// requested DDL cannot be found
+				rows.Close()
+				return "", err
```

Comment:
should return the origin `err` when DDL is not found, but the `err` has been overwritten